### PR TITLE
pass model directory as input to torchserve

### PR DESCRIPTION
**Pull request by mseth10**
*Issue #, if available:*
https://github.com/aws/sagemaker-pytorch-inference-toolkit/issues/117

*Description of changes:*
Removing the [_adapt_to_ts_format](https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/d3fd2f33028a223d836de3f97f7ff429b6d74d00/src/sagemaker_pytorch_serving_container/torchserve.py#L110) function and its test. It's not needed as torchserve accepts model directory as input now.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/sagemaker_pytorch_serving_container/torchserve.py [hunk 1]
```diff
@@ -107,30 +103,6 @@ def start_torchserve(handler_service=DEFAULT_HANDLER_SERVICE):
     ts_process.wait()
 
 
-def _adapt_to_ts_format(handler_service):
-    if not os.path.exists(DEFAULT_TS_MODEL_DIRECTORY):
-        os.makedirs(DEFAULT_TS_MODEL_DIRECTORY)
```
Comment (fm1ch4) [inline src/sagemaker_pytorch_serving_container/torchserve.py:112]: Is this not needed even though we set `--model-store MODEL_STORE` in the torchserve command?

File: src/sagemaker_pytorch_serving_container/torchserve.py [hunk 2]
```diff
@@ -92,7 +88,7 @@ def start_torchserve(handler_service=DEFAULT_HANDLER_SERVICE):
         "--log-config",
         DEFAULT_TS_LOG_FILE,
         "--models",
-        "model.mar"
+        DEFAULT_TS_MODEL_NAME=environment.model_dir
```
Comment (fm1ch4) [inline src/sagemaker_pytorch_serving_container/torchserve.py:91]: Does this need to be `DEFAULT_TS_MODEL_NAME + "=" + environment.model_dir`?

File: src/sagemaker_pytorch_serving_container/torchserve.py (hunk 1 above)
Comment (mseth10) [reply src/sagemaker_pytorch_serving_container/torchserve.py:112]: addressed

File: src/sagemaker_pytorch_serving_container/torchserve.py (hunk 2 above)
Comment (mseth10) [reply src/sagemaker_pytorch_serving_container/torchserve.py]: addressed

**Comment by mseth10** [comment]
Error logs while running https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/integration/local/test_serving.py:
```
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,605 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - Listening on port: /home/model-server/tmp/.ts.sock.9068
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,605 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - [PID]49640
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,605 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - Torch worker started.
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,605 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - Python runtime: 3.8.10
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,605 [INFO ] W-9068-model_1.0 org.pytorch.serve.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.ts.sock.9068
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,606 [INFO ] W-9068-model_1.0 org.pytorch.serve.wlm.WorkerThread - Flushing req. to backend at: 1648863129606
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,606 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - Connection accepted: /home/model-server/tmp/.ts.sock.9068.
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,606 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - model_name: model, batchSize: 1
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - Backend worker process died.
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - Traceback (most recent call last):
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/site-packages/ts/model_loader.py", line 83, in load
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     module, function_name = self._load_handler_file(handler)
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/site-packages/ts/model_loader.py", line 123, in _load_handler_file
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     module = importlib.import_module(module_name)
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/importlib/__init__.py", line 127, in import_module
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     return _bootstrap._gcd_import(name[level:], package, level)
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - ModuleNotFoundError: No module named 'model'
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - 
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - During handling of the above exception, another exception occurred:
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] epollEventLoopGroup-5-71 org.pytorch.serve.wlm.WorkerThread - 9068 Worker disconnected. WORKER_STARTED
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - 
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - Traceback (most recent call last):
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/site-packages/ts/model_service_worker.py", line 189, in <module>
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     worker.run_server()
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/site-packages/ts/model_service_worker.py", line 161, in run_server
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     self.handle_connection(cl_socket)
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/site-packages/ts/model_service_worker.py", line 123, in handle_connection
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     service, result, code = self.load_model(msg)
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/site-packages/ts/model_service_worker.py", line 95, in load_model
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     service = model_loader.load(model_name, model_dir, handler, gpu,
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [WARN ] W-9068-model_1.0 org.pytorch.serve.wlm.BatchAggregator - Load model failed: model, error: Worker died.
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/site-packages/ts/model_loader.py", line 85, in load
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     module = self._load_default_handler(handler)
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/site-packages/ts/model_loader.py", line 128, in _load_default_handler
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     module = importlib.import_module(module_name, 'ts.torch_handler')
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.8/importlib/__init__.py", line 127, in import_module
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -     return _bootstrap._gcd_import(name[level:], package, level)
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,607 [INFO ] W-9068-model_1.0-stdout MODEL_LOG - ModuleNotFoundError: No module named 'ts.torch_handler./opt/ml/model'
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,608 [WARN ] W-9068-model_1.0 org.pytorch.serve.wlm.WorkerLifeCycle - terminateIOStreams() threadName=W-9068-model_1.0-stderr
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,608 [WARN ] W-9068-model_1.0 org.pytorch.serve.wlm.WorkerLifeCycle - terminateIOStreams() threadName=W-9068-model_1.0-stdout
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,608 [INFO ] W-9068-model_1.0 org.pytorch.serve.wlm.WorkerThread - Retry worker: 9068 in 1597 seconds.
[36malgo-1-s7ww9_1  |[0m 2022-04-02T01:32:09,672 [INFO ] W-9044-model_1.0-stderr org.pytorch.serve.wlm.WorkerLifeCycle - Stopped Scanner - W-9044-model_1.0-stderr
```

**Comment by mseth10** [comment]
Following tests are now passing:
- flake8, twine, unit tests
- local integration cpu/gpu tests using DLC and generic images
- SM integration cpu/gpu tests using DLC images

SM integration cpu/gpu tests run using "generic" image are currently failing. Here's the error log from cpu test:
```


2022-04-06T09:00:59,852 [INFO ] main org.pytorch.serve.servingsdk.impl.PluginsManager -  Loading snapshot serializer plugin...
--
2022-04-06T09:00:59,876 [INFO ] main org.pytorch.serve.ModelServer - Loading initial models: /opt/ml/model
2022-04-06T09:00:59,880 [WARN ] main org.pytorch.serve.archive.model.ModelArchive - Model archive version is not defined. Please upgrade to torch-model-archiver 0.2.0 or higher
2022-04-06T09:00:59,880 [WARN ] main org.pytorch.serve.archive.model.ModelArchive - Model archive createdOn is not defined. Please upgrade to torch-model-archiver 0.2.0 or higher
2022-04-06T09:00:59,882 [INFO ] main org.pytorch.serve.wlm.ModelManager - Model model loaded.
2022-04-06T09:00:59,894 [INFO ] main org.pytorch.serve.ModelServer - Initialize Inference server with: EpollServerSocketChannel.
2022-04-06T09:00:59,975 [INFO ] main org.pytorch.serve.ModelServer - Inference API bind to: http://0.0.0.0:8080
2022-04-06T09:00:59,976 [INFO ] main org.pytorch.serve.ModelServer - Initialize Metrics server with: EpollServerSocketChannel.
2022-04-06T09:00:59,977 [INFO ] main org.pytorch.serve.ModelServer - Metrics API bind to: http://127.0.0.1:8082
Model server started.
2022-04-06T09:01:00,306 [WARN ] pool-3-thread-1 org.pytorch.serve.metrics.MetricCollector - worker pid is not available yet.
2022-04-06T09:01:00,374 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - Listening on port: /home/model-server/tmp/.ts.sock.9000
2022-04-06T09:01:00,375 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - [PID]30
2022-04-06T09:01:00,376 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - Torch worker started.
2022-04-06T09:01:00,378 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - Python runtime: 3.7.11
2022-04-06T09:01:00,383 [INFO ] W-9000-model_1.0 org.pytorch.serve.wlm.WorkerThread - Connecting to: /home/model-server/tmp/.ts.sock.9000
2022-04-06T09:01:00,391 [INFO ] pool-3-thread-1 TS_METRICS - CPUUtilization.Percent:0.0\|#Level:Host\|#hostname:container-0.local,timestamp:1649235660
2022-04-06T09:01:00,393 [INFO ] pool-3-thread-1 TS_METRICS - DiskAvailable.Gigabytes:34.40862274169922\|#Level:Host\|#hostname:container-0.local,timestamp:1649235660
2022-04-06T09:01:00,394 [INFO ] pool-3-thread-1 TS_METRICS - DiskUsage.Gigabytes:16.33287811279297\|#Level:Host\|#hostname:container-0.local,timestamp:1649235660
2022-04-06T09:01:00,395 [INFO ] pool-3-thread-1 TS_METRICS - DiskUtilization.Percent:32.2\|#Level:Host\|#hostname:container-0.local,timestamp:1649235660
2022-04-06T09:01:00,395 [INFO ] pool-3-thread-1 TS_METRICS - MemoryAvailable.Megabytes:6194.03125\|#Level:Host\|#hostname:container-0.local,timestamp:1649235660
2022-04-06T09:01:00,397 [INFO ] pool-3-thread-1 TS_METRICS - MemoryUsed.Megabytes:1015.3203125\|#Level:Host\|#hostname:container-0.local,timestamp:1649235660
2022-04-06T09:01:00,398 [INFO ] pool-3-thread-1 TS_METRICS - MemoryUtilization.Percent:17.1\|#Level:Host\|#hostname:container-0.local,timestamp:1649235660
2022-04-06T09:01:00,399 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - Connection accepted: /home/model-server/tmp/.ts.sock.9000.
2022-04-06T09:01:00,402 [INFO ] W-9000-model_1.0 org.pytorch.serve.wlm.WorkerThread - Flushing req. to backend at: 1649235660402
2022-04-06T09:01:00,449 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - model_name: model, batchSize: 1
2022-04-06T09:01:00,504 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - Backend worker process died.
2022-04-06T09:01:00,504 [INFO ] W-9000-model_1.0-stdout MODEL_LOG - Traceback (most recent call last):
2022-04-06T09:01:00,505 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/ts/model_service_worker.py", line 189, in <module>
2022-04-06T09:01:00,505 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     worker.run_server()
2022-04-06T09:01:00,506 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/ts/model_service_worker.py", line 161, in run_server
2022-04-06T09:01:00,506 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     self.handle_connection(cl_socket)
2022-04-06T09:01:00,507 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/ts/model_service_worker.py", line 123, in handle_connection
2022-04-06T09:01:00,507 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     service, result, code = self.load_model(msg)
2022-04-06T09:01:00,508 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/ts/model_service_worker.py", line 96, in load_model
2022-04-06T09:01:00,508 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     batch_size, envelope, limit_max_image_pixels)
2022-04-06T09:01:00,508 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/ts/model_loader.py", line 112, in load
2022-04-06T09:01:00,507 [INFO ] epollEventLoopGroup-5-1 org.pytorch.serve.wlm.WorkerThread - 9000 Worker disconnected. WORKER_STARTED
2022-04-06T09:01:00,509 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     initialize_fn(service.context)
2022-04-06T09:01:00,509 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/sagemaker_pytorch_serving_container/handler_service.py", line 51, in initialize
2022-04-06T09:01:00,509 [WARN ] W-9000-model_1.0 org.pytorch.serve.wlm.BatchAggregator - Load model failed: model, error: Worker died.
2022-04-06T09:01:00,510 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     super().initialize(context)
2022-04-06T09:01:00,510 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/sagemaker_inference/default_handler_service.py", line 66, in initialize
2022-04-06T09:01:00,510 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     self._service.validate_and_initialize(model_dir=model_dir)
2022-04-06T09:01:00,511 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/sagemaker_inference/transformer.py", line 157, in validate_and_initialize
2022-04-06T09:01:00,511 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     self._validate_user_module_and_set_functions()
2022-04-06T09:01:00,511 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/site-packages/sagemaker_inference/transformer.py", line 170, in _validate_user_module_and_set_functions
2022-04-06T09:01:00,512 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     user_module = importlib.import_module(user_module_name)
2022-04-06T09:01:00,512 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/conda/lib/python3.7/importlib/__init__.py", line 127, in import_module
2022-04-06T09:01:00,512 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -     return _bootstrap._gcd_import(name[level:], package, level)
2022-04-06T09:01:00,512 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
2022-04-06T09:01:00,513 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 983, in _find_and_load
2022-04-06T09:01:00,513 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
2022-04-06T09:01:00,513 [WARN ] W-9000-model_1.0 org.pytorch.serve.wlm.WorkerLifeCycle - terminateIOStreams() threadName=W-9000-model_1.0-stderr
2022-04-06T09:01:00,513 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
2022-04-06T09:01:00,514 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap_external>", line 728, in exec_module
2022-04-06T09:01:00,514 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
2022-04-06T09:01:00,514 [WARN ] W-9000-model_1.0 org.pytorch.serve.wlm.WorkerLifeCycle - terminateIOStreams() threadName=W-9000-model_1.0-stdout
2022-04-06T09:01:00,515 [INFO ] W-9000-model_1.0-stdout MODEL_LOG -   File "/opt/ml/model/code/resnet18.py", line 7, in <module>
2022-04-06T09:01:00,542 [INFO ] W-9000-model_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Stopped Scanner - W-9000-model_1.0-stdout
2022-04-06T09:01:00,542 [INFO ] W-9000-model_1.0-stderr org.pytorch.serve.wlm.WorkerLifeCycle - Stopped Scanner - W-9000-model_1.0-stderr
2022-04-06T09:01:00,543 [INFO ] W-9000-model_1.0 org.pytorch.serve.wlm.WorkerThread - Retry worker: 9000 in 1 seconds.


```

**Review by saimidu — DISMISSED** [review]

**Comment by mseth10** [comment]
The failure with SM integration tests in [test_default_inference.py](https://github.com/aws/sagemaker-pytorch-inference-toolkit/blob/master/test/integration/sagemaker/test_default_inference.py) using generic images turned out to be flaky, tests are passing now.

**Review by fm1ch4 — APPROVED** [review]

**Review by josephevans — APPROVED** [review]

**Review by saimidu — COMMENTED** [review]
Approved, even if it doesn't matter now.

**Pull request merged**